style.md

```diff
@@ -0,0 +1,281 @@
+
+## Basic Principles
+
+Keep in mind that these lectures are targeted at students with (at most!) some self-taught Matlab or Python experience.  Consequently, we want to ensure that the code is clear and simple to focus on the Economics/Math and only expect them to write code using a simplified subset of Julia features.  Some guiding principles:
+
+1. Assume this may be the **first programming language** students learn
+2. Use **compact, script-style** code, organized into functions only when it is natural.  Best practices for writing packages and expository/exploratory/interactive code can be different.
+3. Keep things as **close to the whiteboard math** as possible, including in the code structure and notation
+4. Ensure that all **code can be copied and pasted without modification** into functions for performance and modification without changes to scoping (e.g. no `local` or `global` ever required)
+5. **Avoid type annotations** unless they are required for dispatching
+6. **Avoid creating custom types** unless it is absolutely necessary
+7. Beyond the [Julia Style Guide](https://docs.julialang.org/en/v1/manual/style-guide/), avoid unnecessary whitespace lines and redundant comments
+8. Don't use fancy features and control flow from Julia - unless it makes the code look closer to the math
+9. Avoid both micro-optimizations and coding patterns that pessimize (i.e. poor performance with no benefit in code clarity)
+
+We want users to be able to say _"the code is clearer than Matlab, and even closer to the math"_.  If they want to write their own packages, they will need to learn much more about Julia, but most never will need to.
+
+## Naming Conventions, Comments, etc.
+
+- **Use unicode** where possible in names so that symbols match the math in the document
+- **Be careful** about unicode glyphs and symbols which may not be available in the default REPL, Jupyter, etc. for all platforms.  **TODO** 
+- **Do not** use extra whitespace, use comment headers, or redundant comments.  For example, **do not**
+```julia
+# BAD!
+foo(a) #Calls the foo function
+
+# == Parameters == #
+
+bar = 2.0 
+
+# GOOD!
+foo(a)
+
+# Parameters
+bar = 2.0
+```
+- **Do not** align the `=` sign for construction of variables (though acceptable for matrices).  i.e.
+```julia
+# BAD!
+var1       =  1.0
+variable2  =  2.0
+
+# GOOD!
+var1 = 1.0
+variable2 = 2.0
+
+# ACCEPTABLE BUT OFTEN UNNECESSARY
+A = [1 2;
+     3 4]
+```
+- **Do not** use docstrings in any lecture code - except when explaining packages and the `?` environment.
+- **Feel free** to use the `⋅` unicode symbol, i.e. `\cdot<TAB>` instead of `dot( , )`
+- **Use `LaTeXStrings.jl`** for all latex literals, i.e. `L"\hat{\alpha}"` instead of `""\$\\hat{\\alpha}\$""`
+- **Prefer** `in` to `∈` 
```

Comment:
Isn't this an exception to the rule?

Reply:
The rule of using unicode?  Maybe, but not necessarily.  The difference is that I am suggesting using unicode for things that match the whiteboard math in the paper.  If you had a set `A` and wanted to see if `a` was in the set,  and the math in the document said `a ∈ A` then I think it would make sense to use exactly that in the code.

However, when thinking about for loops and comprehensions, it is code rather than math.  I don't have very strong feelings.  The `∈` just seems excessivly fancy to me (kind of like when people you a ∀  symbol in a paper when "for all" would have read perfectly well.

Reply:
I would rather keep it (pro  ∈). At least in my math classes no one would ever write "for all" and use ∀, or such as ∋ or subject to `:` or given `|`. It reads for each element in that collection or for each member of... From a coding/reading it saves a character length in the line.

How about `≠`, `∉`?

Reply:
My issue is that in a construction such as
```julia
for i ∈ 1:10
  @show(i)
end
```
it is inherently code, not the for loop isn't really math.  If it was implementing an actual equation, then I would be for the symbol.  Same with ≠, ∉

The other issue is that students may get confused on how to type it, especially those ones.

But let me ask a better quetsion:  What do the main julia docs do?  If they use ∈ instead of `in` and `≠` instead of `!=`  ... in code blocks, not in exact mappings to algebra of an equations... then I am all for going fancy.

Reply:
No stance. They are aliases so they are the same operator. Biggest difference is probably,
```
julia> in.(1, 1:3)
3-element BitArray{1}:
  true
 false
 false

julia> .∈(1, 1:3)
3-element BitArray{1}:
  true
 false
 false
```

Reply:
@sglyon your thoughts? I don't have very strong opinions, but lean towards `in` for non math code

Reply:
I am in favor of `in` rather than `∈` when writing loops or other code based statements but am also in favor to using `∈` when it allows code to look more like whiteboard math.

Reply:
@Nosferican Sorry, I think you are outvoted on this one.  Things that are inherently code and control flow should stay ascii.  So `in` and `!=`  instead of `∈, ≠`.

I added a point about using those unicode symbols for set inclusion.